pkg/operator/csi/csidrivercontrollerservicecontroller/helpers.go

```diff
@@ -63,3 +65,20 @@ func WithSecretHashAnnotationHook(
 		return nil
 	}
 }
+
+// WithStrategyHook sets the deployment.Spec.Strategy to Recreate on single-node deployments.
+// This is required to avoid deployment rollouts getting stuck on single-node clusters.
+func WithStrategyHook(configinformers configinformers.SharedInformerFactory) DeploymentHookFunc {
+	return func(_ *opv1.OperatorSpec, deployment *appsv1.Deployment) error {
+		infra, err := configinformers.Config().V1().Infrastructures().Lister().Get(infraConfigName)
+		if err != nil {
+			return err
+		}
+		if infra.Status.ControlPlaneTopology == v1.SingleReplicaTopologyMode {
+			deployment.Spec.Strategy = appsv1.DeploymentStrategy{
+				Type: appsv1.RecreateDeploymentStrategyType,
+			}
```

Comment:
We are trying to prevent the deployment rollout getting stuck in single-node clusters. Is setting `deployment.Spec.Strategy=Recreate` the best approach so address this issue?

Or should we [make it RollingUpdate like openshift-oauth-apiserver](https://github.com/openshift/cluster-authentication-operator/blob/master/bindata/oauth-apiserver/deploy.yaml#L12-L20) and set the number of replicas according to the number of master nodes?

@deads2k

Reply:
> Or should we [make it RollingUpdate like openshift-oauth-apiserver](https://github.com/openshift/cluster-authentication-operator/blob/master/bindata/oauth-apiserver/deploy.yaml#L12-L20) and set the number of replicas according to the number of master nodes?
> 
> @deads2k

I prefer the rolling update (if it semantically works for you) and number of replicas.  Now we had to limit surge because of our anti-affinity rules and our interaction with hosts.  if you don't have those contraints, you could choose to surge.

Reply:
It does not work, since most CSI drivers use `hostNetwork: true` and expose ports for metrics & liveness probe. With hostNetwork, these ports become node ports. If we have one pod on each node, a new pod created by rolling upgrade can't find a node with free ports. We will need to use `Recreate` everywhere.

Reply:
@jsafrane I think that's not true when we set `RollingUpdate` with `maxSurge=0`. Essentially this config should allow the new pod to start only when the old one is stopped.

Reply:
Ack, tested `maxSurge: 0` and the Deployment  can proceed even with rolling update + single schedulable node.